src/irmin/contents.ml

```diff
@@ -32,6 +32,139 @@ module Cstruct = struct
   let of_string s = Ok (Cstruct.of_string s)
 end
 
+type json = [
+  | `Null
+  | `Bool of bool
+  | `String of string
+  | `Float of float
+  | `O of (string * json) list
+  | `A of json list
+]
+
+let json =
+  let open Type in
+  mu (fun ty ->
+  variant "json" (fun null bool string float obj arr -> function
+    | `Null -> null
+    | `Bool b -> bool b
+    | `String s -> string s
+    | `Float f -> float f
+    | `O o -> obj o
+    | `A a -> arr a)
+  |~ case0 "null" `Null
+  |~ case1 "bool" bool (fun x -> `Bool x)
+  |~ case1 "string" string (fun x -> `String x)
+  |~ case1 "float" float (fun x -> `Float x)
+  |~ case1 "object" (list (pair string ty)) (fun obj -> `O obj)
+  |~ case1 "array" (list ty) (fun arr -> `A arr)
+  |> sealv)
+
+module Json = struct
+  type t = (string * json) list
+
+  let t = Type.(list (pair string json))
+
+  let merge_object a b =
+    try
+      List.fold_right (fun (k, v) acc ->
+        match List.assoc_opt k b with
+        | Some x when v <> x -> failwith "Unable to merge JSON objects"
+        | Some _ -> acc
+        | _ -> (k, v) :: acc
+      ) a b
+      |> Merge.ok
+    with
+      Failure msg -> Merge.conflict "%s" msg
+
+  let merge ~old t1 t2 =
+    let open Merge.Infix in
+    old () >>=* function
+      | Some j ->
+          if j = t1 then
+            merge_object t1 t2
+          else if j = t2 then
+            merge_object t2 t1
+          else
+            merge_object j t1 >>=* fun t1 ->
+            merge_object t1 t2
```

Comment:
here you probably want `merge_object j t1 t2`: deciding if a value has been updated in either `t1` or `t2` can be done by checking if the association from `j` has changed. A conflict should be raised only if such a value has been updated in both branches.